crates/uv-trampoline-builder/src/lib.rs

```diff
@@ -30,28 +31,160 @@ const LAUNCHER_AARCH64_GUI: &[u8] =
 const LAUNCHER_AARCH64_CONSOLE: &[u8] =
     include_bytes!("../../uv-trampoline/trampolines/uv-trampoline-aarch64-console.exe");
 
+// See `uv-trampoline::bounce`. These numbers must match.
+const PATH_LENGTH_SIZE: usize = size_of::<u32>();
+const MAX_PATH_LENGTH: u32 = 32 * 1024;
+const MAGIC_NUMBER_SIZE: usize = 4;
+
+#[derive(Debug)]
+pub struct Launcher {
+    pub kind: LauncherKind,
+    pub python_path: PathBuf,
+}
+
+impl Launcher {
+    /// Read [`Launcher`] metadata from a trampoline executable file.
+    ///
+    /// Returns `Ok(None)` if the file is not a trampoline executable.
+    /// Returns `Err` if the file looks like a trampoline executable but is formatted incorrectly.
+    ///
+    /// Expects the following metadata to be at the end of the file:
+    ///
+    /// ```text
+    /// - file path (no greater than 32KB)
+    /// - file path length (u32)
+    /// - magic number(4 bytes)
+    /// ```
+    ///
+    /// This should only be used on Windows, but should just return `Ok(None)` on other platforms.
+    ///
+    /// This is an implementation of [`uv-trampoline::bounce::read_trampoline_metadata`] that
+    /// returns errors instead of panicking. Unlike the utility there, we don't assume that the
+    /// file we are reading is a trampoline.
+    #[allow(clippy::cast_possible_wrap)]
+    pub fn try_from_path(path: &Path) -> Result<Option<Self>, Error> {
+        let mut file = File::open(path)?;
+
+        // Read the magic number
+        let Some(kind) = LauncherKind::try_from_file(&mut file)? else {
+            return Ok(None);
+        };
+
+        // Seek to the start of the path length.
+        let Ok(_) = file.seek(io::SeekFrom::End(
+            -((MAGIC_NUMBER_SIZE + PATH_LENGTH_SIZE) as i64),
+        )) else {
+            return Err(Error::InvalidLauncher(
+                "Unable to seek to the start of the path length".to_string(),
+            ));
+        };
+
+        // Read the path length
+        let mut buffer = [0; PATH_LENGTH_SIZE];
+        file.read_exact(&mut buffer)
+            .map_err(|err| Error::InvalidLauncherRead("path length".to_string(), err))?;
+
+        let path_length = {
+            let raw_length = u32::from_le_bytes(buffer);
+
+            if raw_length > MAX_PATH_LENGTH {
+                return Err(Error::InvalidLauncher(format!(
+                    "Only paths with a length up to 32KBs are supported but the Python executable path has a length of {raw_length}"
```

Comment:
nit: move the description into the `thiserror` `error` attribute

Reply:
There are a bunch of these and I wasn't sure it was worth the effort of extra `thiserror` variants. Wdyt?

Reply:
I'm always using thiserror: The biggest benefit for me is that moves the error message strings out of the main logic. It also allows reusing the error in another place, and - if needed - allow matching on the error type in upstream. I tend to see the error enum variants as almost free of cost. It's less pronounced here since we're already using an error enum (and obviously it makes sense to often pass string snippets into error variants as we do a lot), this is mostly out of experience in maturin where it's all anyhow and `.context()` and i wished i had used thiserror consistently.

Reply:
Yeah the nuanced part here is that I don't think you'd ever want to match on different error cases here, i.e., it's more work for a consumer to match on the errors and they'll have to learn that they need to combine a bunch of these cases when they're semantically equivalent to a caller, but I'll try to move some more into the error variant.

Reply:
Actually really satisfying to rewrite these.. lol